slack.c

```diff
@@ -163,26 +164,32 @@ static void slack_login(PurpleAccount *account) {
 				gc->prpl, PURPLE_CALLBACK(slack_conversation_updated), NULL);
 	}
 
-	const gchar *token = purple_account_get_string(account, "api_token", NULL);
-	if (!token || !*token) {
-		purple_debug_warning("slack", "api_token not set; using password as token\n");
-		token = purple_account_get_password(account);
-	}
-	if (!token || !*token) {
+	const gchar *host = g_strrstr(purple_account_get_username(account), "%");
+	if (!host || !*host) {
 		purple_connection_error_reason(gc,
-			PURPLE_CONNECTION_ERROR_INVALID_SETTINGS, "API token required");
+			PURPLE_CONNECTION_ERROR_INVALID_SETTINGS, "Host setting is required");
 		return;
 	}
+	/* move the host pointer forward one character as it is at a % right now */
+	host++;
 
 	SlackAccount *sa = g_new0(SlackAccount, 1);
 	gc->proto_data = sa;
 	sa->account = account;
 	sa->gc = gc;
+	sa->host = g_strdup(host);
 
-	const char *host = strrchr(account->username, '@');
-	sa->api_url = g_strdup_printf("https://%s/api", host ? host+1 : "slack.com");
+	/* check if we have a token and set it as the password if we do */
+	const gchar *token = purple_account_get_string(sa->account, "api_token", NULL);
+	if (token && *token) {
+		sa->token = g_strdup(token);
+		purple_account_set_password(sa->account, sa->token);
```

Comment:
I think set_password does a strdup itself, so this token is leaked in this case (since token is re-assigned later).  Just `purple_account_set_password(sa->account, token)` without assigning `sa->token` might work.

Reply:
`sa->token` is used directly in the api client, so that either needs to be moved to `purple_account_get_password` or we can just keep the token around.  I would opt to keep the token, as in purple3 `purple_account_get_password` has to look up the password from the keyring every time as there is no cache (I'm in the process of removing said cache).